lib/urlwatch/reporters.py

```diff
@@ -201,49 +193,96 @@ def _parts(self):
-            else:
-                yield SafeHtml('<span class="unified_nor">{line}</span>').format(line=line)
+        yield SafeHtml(
+            '<div style="font-family:monospace;font-style:italic">\n'
+            '<span>{pkgname} {version}, {copyright}</span>\n'
+            '<address>Website: <a href="{url}">{url}</a></address>\n'
+            '<span>Watched {count} URLs in {duration} seconds</span>\n'
+            '</div>\n'
+            '</body>\n'
+            '</html>\n').format(pkgname=urlwatch.pkgname, version=urlwatch.__version__, copyright=urlwatch.__copyright__,
+                                url=urlwatch.__url__, count=len(self.job_states), duration=self.duration.seconds)
+
+    def _diff_to_html(self, unified_diff, is_markdown=False):
+        if is_markdown and Markdown:
+            # rebuild html from markdown using markdonw2 library's Markdown
+            markdowner = Markdown(safe_mode='escape', extras=['strike', 'target-blank-links'])
+            atags = re.compile(r'<a ')
+            htags = re.compile(r'<(/?)h\d>')
+            tags = re.compile(r'^<p>(<code>)?|(</code>)?</p>$')
+
+            def mark_to_html(text):
+                '''converts Markdown (as generated by pyhtml2text filter) back to html'''
+                if text == '* * *':  # manually expand horizontal ruler since <hr> tag is used by us to separate jobs
+                    return '-' * 80
+                pre = ''
+                if text.lstrip().startswith('* '):  # item of unordered list
+                    lstripped = text.lstrip()
+                    indent = len(text) - len(lstripped)
+                    pre = '&nbsp;' * indent
+                    pre += '● ' if indent == 2 else '◾ ' if indent == 4 else '○ '
+                    text = text.split('* ', 1)[1]
+                elif text.startswith(' '):  # replace leading spaces or converter will strip
+                    lstripped = text.lstrip()
+                    text = '&nbsp;' * (len(text) - len(lstripped)) + lstripped
+                html_out = markdowner.convert(text).strip('\n')  # convert markdown to html
+                html_out = atags.sub('<a style="font-family:inherit;color:inherit" ', html_out)  # fix <a> tag styling
+                html_out, sub = tags.subn('', html_out)  # remove added tags
+                if sub:
+                    return pre + html_out
+                html_out = htags.sub(r'<\g<1>strong>', html_out)  # replace heading tags with <strong>
+                return pre + html_out
+
+            for line in unified_diff.splitlines():
+                if line[0] == '+':
+                    yield '<span style="color:green">' + '+' + mark_to_html(line[1:]) + '</span>'
+                elif line[0] == '-':
+                    yield '<span style="color:red">' + '-' + mark_to_html(line[1:]) + '</span>'
+                elif line[0] == '@':
+                    yield '<span style="background-color:whitesmoke">' + line + '</span>'
```

Comment:
Please use e.g.:

```
'<span style="color:green">+{html}</span>'.format(html=mark_to_html(line[1:]))
```

Reply:
Also, are you sure this works properly with HTML escaping from the markdown all the time?

Reply:
`str.format()` and later literal string interpolation (aka f-strings, since Python 3.6) are better to read and maintain. Less `'` and `+` and if the string to be inserted is a big expression, the expression is named and the place where it is inserted is easier to see (e.g. `{html}` instead of `' + mark_to_html(line[1:])) + '`).

Also it's easier to make it customizable -- you could let the user provide a string and document which keys are provided without having to modify the code, and the user can move the `{html}` around or even use it multiple times (e.g. `<strong title="{html}">Added line: {html}</strong>`). Not that we should make it customizable, but using strings + `.format()` makes this super easy also in other parts of the code.

Reply:
Thanks for taking the time for the explanation. 

I do personally use f-strings and had to revert to using '+' (now `str.format()`) to testing for 3.5.